connectivity/drivers/cellular/Samsung/TARGET_S5JS100/RIL/src/SAMSUNG_S5JS100_L3IP_Driver.cpp

```diff
@@ -0,0 +1,182 @@
+#include "SAMSUNG_S5JS100_L3IP_Driver.h"
+
+
+#define TRACE_GROUP             "S5JS100_L3IP"
+#define S5JS100_BUFF_ALIGNMENT  0
+
+
+SAMSUNG_S5JS100_L3IP_Driver::SAMSUNG_S5JS100_L3IP_Driver(const char *ifname, int mtu)
+    : _io_device(nullptr),
+      _mtu(mtu)
+{
+    tr_info("IF = %s, MTU = %d", ifname, mtu);
+    strncpy(_ifname, ifname, sizeof(_ifname));
+}
+
+SAMSUNG_S5JS100_L3IP_Driver::~SAMSUNG_S5JS100_L3IP_Driver()
+{
+    power_down();
+}
+
+bool SAMSUNG_S5JS100_L3IP_Driver::link_out(net_stack_mem_buf_t *buf)
+{
+    bool result = true;
+
+    MBED_ASSERT(_memory_manager);
+    MBED_ASSERT(_io_device);
+
+    // Allocate continuous buffer for data
+    const int total_len = _memory_manager->get_total_len(buf);
+    net_stack_mem_buf_t *cont_buf = _memory_manager->alloc_heap(total_len, S5JS100_BUFF_ALIGNMENT);
+
+    if (cont_buf) {
+        uint8_t *cont_buf_ptr = (uint8_t *)_memory_manager->get_ptr(cont_buf);
+
+        /* copy all frames to continuous buffer */
+        for (net_stack_mem_buf_t *q = buf; q != nullptr; q = _memory_manager->get_next(q)) {
+            const uint8_t *ptr = (const uint8_t *)_memory_manager->get_ptr(q);
+            const uint16_t len = _memory_manager->get_len(q);
+
+            memcpy(cont_buf_ptr, ptr, len);
+            cont_buf_ptr += len;
+        }
+
+        const int sent = _io_device->write((const char *)_memory_manager->get_ptr(cont_buf), total_len);
+        tr_debug("link_out (%s): sent %d/%d bytes", _ifname, sent, total_len);
+        if (sent != total_len) {
+            result = false;
+        }
+
+        _memory_manager->free(cont_buf);
+    } else {
+        tr_error("Unable to allocate buffer for TX data!");
+        result = false;
+    }
+
+    _memory_manager->free(buf);
+
+    return result;
+}
+
+bool SAMSUNG_S5JS100_L3IP_Driver::power_up()
+{
+    _io_device = getModemIoDeviceByName(_ifname);
+    tr_debug("power_up: io_device = %p", _io_device);
+    if (_io_device) {
+        _io_device->register_ReadCb(read_cb, this);
+        _l3ip_link_state_cb(true);
+        return true;
+    } else {
+        tr_error("Unable to get modem IO device!");
+        return false;
+    }
+}
+
+uint32_t SAMSUNG_S5JS100_L3IP_Driver::get_mtu_size() const
+{
+    tr_debug("get_mtu_size: mtu = %d", _mtu);
+    return _mtu;
+}
+
+uint32_t SAMSUNG_S5JS100_L3IP_Driver::get_align_preference() const
+{
+    tr_debug("get_align_preference: align = %d", S5JS100_BUFF_ALIGNMENT);
+    return S5JS100_BUFF_ALIGNMENT;
+}
+
+void SAMSUNG_S5JS100_L3IP_Driver::get_ifname(char *name, uint8_t size) const
+{
+    strncpy(name, _ifname, size);
+}
+
+void SAMSUNG_S5JS100_L3IP_Driver::set_link_input_cb(l3ip_link_input_cb_t input_cb)
+{
+    _l3ip_link_input_cb = input_cb;
+}
+
+void SAMSUNG_S5JS100_L3IP_Driver::set_link_state_cb(l3ip_link_state_change_cb_t state_cb)
+{
+    _l3ip_link_state_cb = state_cb;
+}
+
+void SAMSUNG_S5JS100_L3IP_Driver::add_ipv4_multicast_group(const SocketAddress &address)
```

Comment:
why are these emtpy functions ?

Reply:
`class SAMSUNG_S5JS100_L3IP_Driver : public L3IP`. 
In `L3IP` class, these functions is virtual. So in `SAMSUNG_S5JS100_L3IP_Driver` class we need to implement them, but these functions are not used in our current logic, so the implementation is empty. If there is any expansion in the future, we will continue to implement them

Reply:
Can you drop a comment to the top function with reasoning why these are empty (it will be good to know).

I assume these are not mandatory for cellular functionality or provided somewhere else (isn't there default implementation?) or these must be defined here otherwise an error?

Reply:
> Can you drop a comment to the top function with reasoning why these are empty (it will be good to know).

OK, we will add a comment to function.

> I assume these are not mandatory for cellular functionality or provided somewhere else (isn't there default implementation?) or these must be defined here otherwise an error?

Yes, these must be defined here, if not defined, the compile will error. Because these functions is virtual (` virtual void add_ipv4_multicast_group(const SocketAddress &address) = 0;)` in the base class `L3IP`, `SAMSUNG_S5JS100_L3IP_Driver` class is a derived class, so must to implement these. 
The default implementation is not given in the mbed-os souce. So we No-op at this stage.